caddyhttp/proxy/reverseproxy.go

```diff
@@ -214,6 +220,9 @@ func (p *ReverseProxy) ServeHTTP(rw http.ResponseWriter, outreq *http.Request, r
 }
 
 func (p *ReverseProxy) copyResponse(dst io.Writer, src io.Reader) {
+	buf := bufferPool.Get()
+	defer bufferPool.Put(buf)
```

Comment:
@mholt I find this during the memory profiling, here it is:

```
       4MB   917.01MB (flat, cum) 81.12% of Total
         .          .    217:   if p.FlushInterval != 0 {
         .          .    218:           if wf, ok := dst.(writeFlusher); ok {
         .          .    219:                   mlw := &maxLatencyWriter{
         .          .    220:                           dst:     wf,
         .          .    221:                           latency: p.FlushInterval,
       4MB        4MB    222:                           done:    make(chan bool),
         .          .    223:                   }
         .          .    224:                   go mlw.flushLoop()
         .          .    225:                   defer mlw.stop()
         .          .    226:                   dst = mlw
         .          .    227:           }
         .          .    228:   }
         .   913.01MB    229:   io.Copy(dst, src)
         .          .    230:}
```

As you see `io.Copy` in `proxy.(*ReverseProxy).copyResponse` allocated a lot of memory and the culprit is:

```
 912.43MB   913.01MB (flat, cum) 80.77% of Total
         .          .    382:   // Similarly, if the writer has a ReadFrom method, use it to do the copy.
         .          .    383:   if rt, ok := dst.(ReaderFrom); ok {
         .          .    384:           return rt.ReadFrom(src)
         .          .    385:   }
         .          .    386:   if buf == nil {
  912.43MB   912.43MB    387:           buf = make([]byte, 32*1024)
         .          .    388:   }
         .          .    389:   for {
         .          .    390:           nr, er := src.Read(buf)
         .          .    391:           if nr > 0 {
         .   596.16kB    392:                   nw, ew := dst.Write(buf[0:nr])
         .          .    393:                   if nw > 0 {
         .          .    394:                           written += int64(nw)
         .          .    395:                   }
         .          .    396:                   if ew != nil {
         .          .    397:                           err = ew
```

It allocate a 32k buffer for every copying(in other word, every request) because we did not supply the buffer, so I  think we could use buffer pool to offer and recycle the temporary buffer.
And it seems works well:

```
ROUTINE ======================== github.com/mholt/caddy/caddyhttp/proxy.(*ReverseProxy).copyResponse in /home/tw/golib/src/github.com/mholt/caddy/caddyhttp/proxy/reverseproxy.go
       4MB    19.05MB (flat, cum)  8.37% of Total
         .          .    218:
         .          .    219:   return nil
         .          .    220:}
         .          .    221:
         .          .    222:func (p *ReverseProxy) copyResponse(dst io.Writer, src io.Reader) {
         .    13.93MB    223:   buf := bufferPool.Get()
         .          .    224:   defer bufferPool.Put(buf)
         .          .    225:
         .          .    226:   if p.FlushInterval != 0 {
         .          .    227:           if wf, ok := dst.(writeFlusher); ok {
         .          .    228:                   mlw := &maxLatencyWriter{
         .          .    229:                           dst:     wf,
         .          .    230:                           latency: p.FlushInterval,
       4MB        4MB    231:                           done:    make(chan bool),
         .          .    232:                   }
         .          .    233:                   go mlw.flushLoop()
         .          .    234:                   defer mlw.stop()
         .          .    235:                   dst = mlw
         .          .    236:           }
         .          .    237:   }
         .     1.12MB    238:   io.CopyBuffer(dst, src, buf.([]byte))
         .          .    239:}
         .          .    240:
         .          .    241:type writeFlusher interface {
         .          .    242:   io.Writer
         .          .    243:   http.Flusher

```